### PR TITLE
Use __invoke for InvocationStrategyInterface and remove container

### PR DESCRIPTION
**Pull request by opengeek**
Refactored InvocationStrategyInterface and implementation; now with working code and tests to prove it for both included strategies!

### CODE REVIEW

File: Slim/App.php [hunk 1]
```diff
@@ -371,9 +371,12 @@ public function __invoke(ServerRequestInterface $request, ResponseInterface $res
...
         $routeInfo = $this->container->get('router')->dispatch($request);
         if ($routeInfo[0] === Dispatcher::FOUND) {
-            /** @var InvocationStrategyInterface $handler */
-            $handler = $this->container->get('foundHandler');
-            return $handler->invoke($this->container, $routeInfo, $request, $response);
+            $routeArguments = [];
+            foreach ($routeInfo[2] as $k => $v) {
+                $routeArguments[$k] = urldecode($v);
+            }
+            $request = $request->withAttribute('routeArguments', $routeArguments);
```
Comment (akrabat) [inline Slim/App.php:378]: Why have a sub-array for these arguments? It now means that you can no longer add an attribute in app middleware and have it available in the callback's `$args`.

Comment (akrabat) [inline Slim/App.php:378]: It also means that there's no way to do default arguments with a stock install of Slim3 without using `isset`.

Comment (opengeek) [inline Slim/App.php:378]: The isolation was at Josh's suggestion as attributes really aren't the right place for storing route arguments. Default arguments can be done—even in the default RequestResponse strategy—see latest commits on #1343.

Comment (silentworks) [inline Slim/App.php:378]: @opengeek take a look at #1348 as the currently merged changes broke the default setup, we think this will remedy it.

**Pull request merged**